### ISSUE TITLE
Issues to solve with locking

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#16664]
- [x] TODO: lock scanner: prevent the scanner to scan or modify/delete locked elements (see https://github.com/owncloud/core/issues/13391#issuecomment-105852842). Maybe the scanner should also use acquireLock to make sure other processes do not try and do something with the cache entries that are about to be changed. This also means that concurrent scanner processes would exclude each other.
- [x] TODO: also acquire shared lock before rename pre-hooks (as we did in basicOperation): https://github.com/owncloud/core/issues/16829
- [ ] TODO: wrap touch operation from upload (chunking mode) in the transaction: https://github.com/owncloud/core/issues/13498 (non-chunking should be fine). This will also need the "scanner locking" to avoid the scanner finding the non-touched-file with a different timestamp. => PR here https://github.com/owncloud/core/pull/17104
- [x] TODO: also properly wrap $view->file_put_contents with locks. I see it calls $view->fopen() but the cache update is done after fclose(). Might be safer to also include the cache update and post hooks inside a transaction/lock. https://github.com/owncloud/core/issues/16727

- [x] BUG: rename folder during upload still has inconsistency between cache and on-disk data (https://github.com/owncloud/core/issues/11795)

0. Add `sleep(10);` here: https://github.com/owncloud/core/blob/master/lib/private/files/cache/cache.php#L551
1. Setup sync client
2. Close sync client
3. Copy a big folder "test" locally
4. Start the sync client
5. Quickly, while it starts uploading: refresh the web UI
6. Rename "test" to "re"
7. Sync client fails with error "could not find folder 'test'" after a bunch of uploads, which is expected
8. In web UI: "re" is still empty, but "data/$user/re" contains files
9. Sync client resumes upload, but in the end not all files are present in "test"

We need to tweak the locking to be able to avoid such situations.

- [x] rename folder with slow cache update still loses contents, see https://github.com/owncloud/core/issues/13391#issuecomment-105525369
- [x] BUG: properly handle LockedException in web UI: when renaming a locked folder in the web UI: exception returned as "500 Internal Server Error" but logged properly:
- [x] BUG: make etag precondition + final rename atomic: https://github.com/owncloud/core/issues/16569 (the current locking did not help, might need to retry after acquireLock is set for small files)
- [x] BUG: sleep after chunk assembly but before final rename, then rename parent folder: sync fails with "could not rename part file to final file". We should bundle chunking inside a transaction. But the current behavior might be ok as it doesn't cause any data corruption.
- [x] BUG: [minor] move big folder to ext storage in web UI, sync client running: sync client cannot download any of the moved files as long as the move is happening (even if the files are there already). Once the move is done, the sync client retries and can get the file. Would still be nice to have to be able to download files as soon as they are moved (but keep the lock on the folder)
To clarify: download should be forbidden in the **source** folder, but not in the **target** folder.

Tests:

- [X] TEST: move big folder to ext storage. During the slow move, cannot rename source or target folder: locked.
- [x] TEST: cannot rename folder while it is being moved to trash (tested with slow ext storage, copy from storage to trash)
- [x] TEST: cannot upload into a folder while it is being moved to trash
- [x] TEST: cannot rename a folder while chunks are being assembled (added sleep() in the chunk assembly loop), LockedException
- [x] TEST: slow down streamCopy for WebDAV upload and try renaming the parent folder
- [X] TEST: delete folder while uploading inside (with slow streamCopy): denied by locked error
- [X] TEST: kill sync client during upload (with slow streamCopy): lock expires after PHP timeout
- [X] TEST: simulate "expected filesize did not match" error: lock was properly released despite exception (shutdown handler)
- [x] TEST: cannot rename target folder when restoring it from trash
- [x] TEST: cannot download files inside a folder that is being moved to trash (source folder)
- [x] TEST: restore big folder from trash to SFTP (slow), try restoring it at the same time. Restore fails due to lock.
- [x] TEST: rename "test" to "test2", sleep before the actual rename: cannot create "test2" in another thread because locked

#### Sharing (locks must work across users):
- [x] TEST: slow upload of small file as owner, recipient cannot download while locked
- [x] TEST: slow upload of small file as recipient, owner cannot download while locked
- [x] TEST: owner moves folder, recipient cannot rename source or target folder
- [x] TEST: recipient moves folder, owner cannot rename source or target folder
- [x] TEST: delete subfolder as recipient, owner cannot rename/download contents
- [x] TEST: delete subfolder as owner, recipient cannot rename/download contents
- [ ] TEST: owner unshares folder (can be slow with encryption), recipient cannot rename/download folder contents => raised as https://github.com/owncloud/core/issues/17244
- [ ] TEST: delete file in thread A, try and share it in thread B. Share must fail because of lock. :no_entry_sign: => raised as https://github.com/owncloud/core/issues/17243
- [x] TEST: delete subfolder as owner in thread A, recipent can still unshare from self in thread B(TBC?) :question: 
- [x] TEST: delete subfolder as recipient in thread A, owner can still unshare from that recipient in thread B (TBC?) :question: works but produces warning (sleep in move2trash after setupTrash)
- [x] TEST: unshare from self twice
- [ ] TEST: share with group/user with encryption enabled (make key creation slow with sleep): recipient should not get access to the files until this is done :no_entry_sign: FAIL, raised as https://github.com/owncloud/core/issues/16621
- [x] TEST: as owner, rename parent folder of shared folder when locked by recipient
- [x] TEST: as recipient, rename parent folder of shared folder when locked by owner

#### Versions:
- [ ] TEST: cannot download version of a file being moved (neither from source nor target, single and cross-storage) :no_entry_sign: FAIL, missing lock (sleep after acquireLock in $view->rename()) => raised as https://github.com/owncloud/core/issues/17245
- [ ] TEST: cannot restore version of a file being moved (neither from source nor target, single and cross-storage). :no_entry_sign: FAIL: restore is blocked by write lock, but the original file gets deleted! (sleep after acquireLock in $view->rename()) => https://github.com/owncloud/core/issues/17245
- [ ] TEST: restore version (with sleep in restore code) in thread A, cannot rename or overwrite the file in thread B => https://github.com/owncloud/core/issues/17269
- [x] TEST: sleep in loop that moves each version (https://github.com/owncloud/core/blob/master/apps/files_versions/lib/storage.php#L270); thread A moves file, thread B overwrites new file name. While the versions are moved (post hook), the file is still locked with a share lock: https://github.com/owncloud/core/pull/17030
- [ ] TEST: sleep in version expiration loop (https://github.com/owncloud/core/blob/master/apps/files_versions/lib/storage.php#L617), trigger cron.php in thread A, try moving the file in thread B :no_entry_sign:

#### Previews:
- [x] TEST: if file was locked (not downloadable) during preview creation, preview needs to retry later
(achieved by overwriting a pic file that has no preview yet, and sleep in streamCopy)

#### Encryption:
- [x] TEST: general testing if locking works like

#### Zip download:
- [x] TEST: download zip of a folder containing a locked new file (sleep after acquireLock before final rename) => new file does not appear in zip
- [ ] TEST: download zip of a folder containing a locked file to be overwritten (sleep after acquireLock before final rename) => zip file is empty :no_entry_sign: FAIL, should skip locked file ?! => https://github.com/owncloud/core/issues/16960

#### s2s and locking
- [ ] TEST: share "local" from server A to server B. On server A, slow-upload+overwrite file (sleep after acquireLock in sabre/file). Server B cannot download. :no_entry_sign: FAIL for some reason server B can still download => raised as https://github.com/owncloud/core/issues/17275

**Comment by PVince81** [comment]
@icewind1991 see above

**Comment by PVince81** [comment]
I suggest to focus on the scanner first so we can get the most critical issue solved at last.

**Comment by PVince81** [comment]
- [ ] BUG: properly show locking error message in the web UI: https://github.com/owncloud/core/issues/16701

**Comment by icewind1991** [comment]
> BUG: rename folder during upload still has inconsistency between cache and on-disk data 

Not really something we can fix without having the client manually acquire a lock on the folder it's uploading to, it's not an issue with concurrent requests

**Comment by PVince81** [comment]
@icewind1991 I still think it should be possible to recover from this situation without having missing files, even if one half is in one folder and the other half in the other one. Goal is to avoid having a situation where files are completely lost. Either the rename operation needs to stop/fail or the sync client will give up once one of the uploads doesn't reach its target.
Not sure, but this might be related to scanner locking too.

**Comment by icewind1991** [comment]
> Not sure, but this might be related to scanner locking too.

It's not

The problem has nothing to do with concurrent requests, it's caused because uploading a folder from the sync client takes multiple requests, there is no way to make that atomic without manually managing a lock for it.

A different way to fix it unrelated to locking would be to have the sync client specify the fileid of the target folder, that way the server can either return a proper error (folder renamed to ...., try again with that path) or get the target folder from that id

**Comment by PVince81** [comment]
The upload doesn't have to be atomic. The sync client would upload 5 files of 10, then get the locked error because someone started renaming the folder (the folder rename must be atomic), and it would give up until the folder is freed again. So in the end the first folder would have 5 files, and then the sync client would recreate the original folder name (once unlocked) to reupload all files.
Hmm I guess it means in the end we'd have a folder with 5 files (the renamed one) and another folder with the original name but all the 10 ones.
But then there's the etag/rename change that might be detected.
Tricky indeed.

**Comment by PVince81** [comment]
I'll try and go through the list again and raise separate tickets.

But before I do we need to fix the regressions: https://github.com/owncloud/core/issues/16998
and also https://github.com/owncloud/core/pull/16963 which has a high risk of influencing a lot of the code paths, which means that retesting only makes sense once that one is finished.

**Comment by cmonteroluque** [comment]
thank you @PVince81

**Comment by PVince81** [comment]
Assigning to self to verify the failed cases I found before.
If they still happen, I'll make separate tickets for them.

I expect the following to still have issues: sharing (share and unshare currently do not set locks) and also versions (version operations aren't transactional yet). Hopefully these should not be critical.

**Comment by PVince81** [comment]
Moving to 8.1.1.

Remaining issues are related to versions that is lacking locking.

**Comment by PVince81** [comment]
All outstanding issues have been raised as separate tickets, closing this one.
